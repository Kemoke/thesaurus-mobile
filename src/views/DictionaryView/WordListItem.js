import React from 'react';
import {FlatList, View} from "react-native";
import {ListItem} from "react-native-elements";

type WordListItemProps = {
    item: {
        word: string,
        synonyms: Array<string>
    }
}

export default class WordListItem extends React.Component<WordListItemProps>{
    state = {
        expanded: false
    };

    renderSynonymItem = ({item}) => (
        <View>
            <ListItem containerStyle={{backgroundColor: '#EEEEEE'}}
                      contentContainerStyle={{paddingLeft: 16}}
                      title={item}
                      bottomDivider/>
        </View>
    );

    toggleSynonymList = () => this.setState({expanded: !this.state.expanded});

    keyExtractor = (item) => item;

    render() {
        const {item} = this.props;
        const {expanded} = this.state;
        return (
            <View>
                <ListItem title={item.word}
                          badge={{value: item.synonyms.length}}
                          bottomDivider
                          chevron
                          testID='wordItem'
                          onPress={this.toggleSynonymList}/>
                {expanded && <FlatList keyExtractor={this.keyExtractor}
                                       data={item.synonyms}
                                       renderItem={this.renderSynonymItem}/>}
            </View>
        );
    }
}