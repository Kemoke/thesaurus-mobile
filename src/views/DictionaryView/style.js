import {StyleSheet} from 'react-native';

export default StyleSheet.create({
    messageContainer: {
        flexGrow: 1,
        alignItems: 'center',
        justifyContent: 'center'
    },
    mainContainer: {
        height: '100%',
        backgroundColor: '#FFF'
    },
    messageText: {
        fontSize: 20,
        color: '#AAAAAA',
        textAlign: 'center',
        marginLeft: 80,
        marginRight: 80
    }
});