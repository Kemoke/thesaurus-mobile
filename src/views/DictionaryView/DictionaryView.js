import React from 'react';
import {FlatList, Text, View} from "react-native";
import {Header, SearchBar} from "react-native-elements";
import styles from './style';
import {SynonymContextWrapper} from "../../SynonymContext";
import type {SynonymContextProps} from "../../SynonymContext";
import WordListItem from "./WordListItem";
import KeyboardAvoidingView from "../../util/KeyboardAvoidingView";

type DictionaryViewProps = {
    synonymContext: SynonymContextProps
}

export class DictionaryViewComponent extends React.Component<DictionaryViewProps> {

    state = {
        search: ''
    };

    updateSearch = search => this.setState({search});

    keyExtractor = (item) => item.word;

    render() {
        const {search} = this.state;
        const wordEntries = Object.entries(this.props.synonymContext.words).map((entry) => ({
            word: entry[0],
            synonyms: entry[1]
        })).filter(entry => entry.word.indexOf(search) !== -1);
        return (
            <KeyboardAvoidingView verticalOffset={40} style={styles.mainContainer}>
                <Header centerComponent={{text: 'DICTIONARY', style: {color: "#FFF"}}}/>
                <SearchBar placeholder="Search for synonyms"
                           lightTheme
                           testID='searchBar'
                           value={this.state.search}
                           onChangeText={this.updateSearch}/>
                {wordEntries.length === 0 ?
                    <View style={styles.messageContainer}>
                        <Text style={styles.messageText}>
                            No synonyms found
                        </Text>
                    </View>
                    :
                    <FlatList keyExtractor={this.keyExtractor}
                              data={wordEntries}
                              testID='wordList'
                              renderItem={({item}) => <WordListItem item={item}/>}/>
                }
            </KeyboardAvoidingView>
        );
    }
}

export default SynonymContextWrapper(DictionaryViewComponent)