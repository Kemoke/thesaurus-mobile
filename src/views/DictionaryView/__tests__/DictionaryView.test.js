import React from 'react';
import {shallow} from "enzyme";
import {DictionaryViewComponent} from "../DictionaryView";

describe('Dictionary view', function () {
    it('should render properly', function () {
        const context = {
            words: {
                "test": ["test2", "test3"],
                "test2": ["test"],
                "test3": ["test"]
            }
        };
        const component = shallow(
            <DictionaryViewComponent synonymContext={context}/>
        );
        expect(component.instance().keyExtractor({word: "test"})).toEqual("test");
        expect(component).toMatchSnapshot();
    });
    it('should render empty properly', function () {
        const context = {
            words: {}
        };
        const component = shallow(
            <DictionaryViewComponent synonymContext={context}/>
        );
        expect(component).toMatchSnapshot();
    });
    it('should render correct list component', function () {
        const context = {
            words: {
                "test": ["test2", "test3"],
                "test2": ["test"],
                "test3": ["test"]
            }
        };
        const component = shallow(
            <DictionaryViewComponent synonymContext={context}/>
        );
        const RenderedComponent = component.findWhere(node => node.prop('testID') === 'wordList').props().renderItem;
        const listComponent = shallow(
            <RenderedComponent item={{
                word: "test",
                synonyms: []
            }}/>
        );
        expect(listComponent).toMatchSnapshot();
    });
    it('should handle search input', function () {
        const context = {
            words: {}
        };
        const component = shallow(
            <DictionaryViewComponent synonymContext={context}/>
        );
        component.findWhere(node => node.prop('testID') === "searchBar").props().onChangeText("test");
        expect(component.state("search")).toEqual("test");
    });
});