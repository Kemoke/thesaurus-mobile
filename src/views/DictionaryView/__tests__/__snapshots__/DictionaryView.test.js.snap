// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dictionary view should render correct list component 1`] = `
<WordListItem
  item={
    Object {
      "synonyms": Array [],
      "word": "test",
    }
  }
/>
`;

exports[`Dictionary view should render empty properly 1`] = `
<KeyboardAvoidingView
  style={
    Object {
      "height": "100%",
    }
  }
  verticalOffset={40}
>
  <Themed.Header
    centerComponent={
      Object {
        "style": Object {
          "color": "#FFF",
        },
        "text": "DICTIONARY",
      }
    }
  />
  <ForwardRef(Themed.SearchBar)
    lightTheme={true}
    onChangeText={[Function]}
    placeholder="Search for synonyms"
    testID="searchBar"
    value=""
  />
  <Component
    style={
      Object {
        "alignItems": "center",
        "flexGrow": 1,
        "justifyContent": "center",
      }
    }
  >
    <Component
      style={
        Object {
          "color": "#AAAAAA",
          "fontSize": 20,
          "marginLeft": 80,
          "marginRight": 80,
          "textAlign": "center",
        }
      }
    >
      No synonyms found
    </Component>
  </Component>
</KeyboardAvoidingView>
`;

exports[`Dictionary view should render properly 1`] = `
<KeyboardAvoidingView
  style={
    Object {
      "height": "100%",
    }
  }
  verticalOffset={40}
>
  <Themed.Header
    centerComponent={
      Object {
        "style": Object {
          "color": "#FFF",
        },
        "text": "DICTIONARY",
      }
    }
  />
  <ForwardRef(Themed.SearchBar)
    lightTheme={true}
    onChangeText={[Function]}
    placeholder="Search for synonyms"
    testID="searchBar"
    value=""
  />
  <FlatList
    data={
      Array [
        Object {
          "synonyms": Array [
            "test2",
            "test3",
          ],
          "word": "test",
        },
        Object {
          "synonyms": Array [
            "test",
          ],
          "word": "test2",
        },
        Object {
          "synonyms": Array [
            "test",
          ],
          "word": "test3",
        },
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReachedThreshold={2}
    renderItem={[Function]}
    scrollEventThrottle={50}
    testID="wordList"
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</KeyboardAvoidingView>
`;
